CM7/Core/Src/gatedriver.c

```diff
@@ -81,10 +90,62 @@ int16_t gatedrv_read_dc_current(gatedriver_t* drv)
 }
 
 /* Note: This has to atomically write to ALL PWM registers */
-//TODO: mechanism for PWM synchronization
-int16_t gatedrv_write_pwm(gatedriver_t* drv)
+int16_t gatedrv_write_pwm(gatedriver_t* drv, float dutyCycles[])
 {
-
+	// computing pulses
+	uint32_t pulses[3];
+	pulses[0] = (uint32_t) (dutyCycles[0] * PERIOD_VALUE / 100);
+	pulses[1] = (uint32_t) (dutyCycles[1] * PERIOD_VALUE / 100);
+	pulses[2] = (uint32_t) (dutyCycles[2] * PERIOD_VALUE / 100);
+
+	// Common configuration for all channels
+	TIM_OC_InitTypeDef PWMConfig;
+	PWMConfig.OCMode       = TIM_OCMODE_PWM1;
+	PWMConfig.OCPolarity   = TIM_OCPOLARITY_HIGH;
+	PWMConfig.OCNPolarity  = TIM_OCNPOLARITY_HIGH;
+	PWMConfig.OCIdleState  = TIM_OCIDLESTATE_SET;
+	PWMConfig.OCNIdleState = TIM_OCNIDLESTATE_RESET;
+	PWMConfig.OCFastMode   = TIM_OCFAST_DISABLE;
+
+	// Attempting to set channel 1
+	PWMConfig.Pulse = pulses[0];
+	if(HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_1) != HAL_OK)
```

Comment:
Kinda a general higher level question, I think that realistically at 480MHz these 3 calls should be very close to eachother but still might not be _fully_ in sync which might cause some stall in the motor when changing duty cycles (which will happen A LOT). Are these synced on a hardware level? I remember you talking about being able to have other channels follow another channel. If we are synchronizing on a hardware level, we should explicitly call out that we are and how

Reply:
I mean, all the PWM outputs follow the same clock, but I don't know if there's any particular way to make them all have changes to their duty cycles go into effect at the same time, if that's what you're asking. I'm not terribly concerned about it though, given the microcontroller clock will run thousands of instructions for every step in the PWM counter, so odds are they'll likely be synchronized anyway.